### PR TITLE
rpb-qt: qt5/eglfs image (no window manager)

### PR DESCRIPTION
**Pull request by ldts**
This pull requests allows the generation of an image suitable for embedded systems not needing a window manager (x11 or Wayland).

It enables Qt5/eglfs (http://doc.qt.io/qt-5/embedded-linux.html) 

The user can validate the native window using Qt5_CinematicExperience as follows:

$ cd /usr/share/cinematicexperience-1.0/
$ ./Qt5_CinematicExperience --platform eglfs

This commit was tested for the db410c release:
IMAGES=rpb-qt5-image
DISTRO=rpb-qt5

### CODE REVIEW

**Comment by koenkooi** [comment]
is this a qt5 distro or an eglfs? It looks to me like it's not really 'rpb-qt5'.

**Comment by ndechesne** [comment]
yes, it is a specific 'distro' to enable eglfs, so maybe rpb-qt5-eglfs might be a better name, but it is a long name, and I don't like it a lot ;-)

I would think that qt5 should work in rpb-x11 and rpb-wayland distro already, but with their respective backend. it might be interesting to try the qt5 image sample on the other distro, it might "just work".

File: conf/distro/rpb-qt5.conf [hunk 1]
```diff
@@ -0,0 +1,8 @@
+require conf/distro/include/rpb.inc
+
+DISTRO_NAME = "Reference-Platform-Build-QT5"
+
+PACKAGECONFIG_append_pn-qtbase = " gles2 eglfs fontconfig kms"
```
Comment (ndechesne) [inline conf/distro/rpb-qt5.conf]: let's add examples as well in PACKAGECONFIG and add them into the image by default.

File: conf/distro/rpb-qt5.conf [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+
+DISTRO_FEATURES_remove = "wayland x11"
+DISTRO_FEATURES_append = " pam systemd opengl"
```
Comment (ndechesne) [inline conf/distro/rpb-qt5.conf]: this is taken care by rpb.inc already. you shouldn't need them here, and should rely on the defaults from rpb.inc

File: recipes-samples/images/rpb-qt5-image.bb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+IMAGE_FEATURES += " package-management debug-tweaks ssh-server-openssh hwcodecs tools-debug"
+
+LICENSE = "MIT"
+
+inherit core-image distro_features_check extrausers
+
+# let's make sure we have a good image..
+REQUIRED_DISTRO_FEATURES = "pam systemd opengl"
+
+# make sure we boot to desktop
```
Comment (ndechesne) [inline recipes-samples/images/rpb-qt5-image.bb]: the comment here is wrong.

Comment (ndechesne) [inline recipes-samples/images/rpb-qt5-image.bb]: and multi-user.target is already the default (i think).

File: recipes-samples/images/rpb-qt5-image.bb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+# by default and without x11-base in IMAGE_FEATURES we default to multi-user.target
+SYSTEMD_DEFAULT_TARGET = "multi-user.target"
+
+CORE_IMAGE_BASE_INSTALL += " \
```
Comment (ndechesne) [inline recipes-samples/images/rpb-qt5-image.bb]: why did you remove gstreamer and gps packages compared to the other images?

File: recipes-samples/images/rpb-qt5-image.bb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    alsa-utils-aplay \
+    networkmanager \
+    networkmanager-nmtui \
+    coreutils \
+    gptfdisk \
+    hostapd \
+    iptables \
+    kernel-modules \
+    sshfs-fuse \
+    cinematicexperience \
```
Comment (ndechesne) [inline recipes-samples/images/rpb-qt5-image.bb]: as sais earlier let's add the qt5 examples package as well

File: recipes-samples/images/rpb-qt5-image.bb (hunk 4 above)
Comment (ndechesne) [inline recipes-samples/images/rpb-qt5-image.bb]: we probably want to keep the x11 packages as well, but check for X11 in DISTRO_FEATURES. I believe this image should be used for DISTRO=rpb and DISTRO=rpb-wayland, and in case of rpb it should be using Qt5 over x11 backend, so the x11 packages would be needed.

in fact we might eventually end up with a single desktop image, who knows..

File: conf/distro/rpb-qt5.conf (hunk 2 above)
Comment (ldts) [inline conf/distro/rpb-qt5.conf]: ok

**Comment by ldts** [comment]
hi @ndechesne, implemented the changes, rebased, tested and pushed.

File: recipes-samples/images/rpb-qt5-image.bb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+CORE_IMAGE_BASE_INSTALL += " \
+    96boards-tools \
+    alsa-utils-aplay \
+    networkmanager \
+    networkmanager-nmtui \
+    coreutils \
+    gps-utils \
+    gpsd \
+    gptfdisk \
+    ${@bb.utils.contains("DISTRO_FEATURES", "x11", "gstreamer1.0-plugins-bad-meta", "", d)} \
```
Comment (ndechesne) [inline recipes-samples/images/rpb-qt5-image.bb]: i think we should not need to check for X11 here. if we need , it looks like a bug in oe-core. did you really need it? if so, do you have a log of the failure without it?

File: recipes-samples/images/rpb-qt5-image.bb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    sshfs-fuse \
+    cinematicexperience \
+    qt5everywheredemo \
+    qt5ledscreen \
+    qt5nmapcarousedemo \
+    qt5nmapper \
+    qt5-opengles2-test \
+    qtsmarthome \
+    quitbattery \
+    quitindicators \
```
Comment (ndechesne) [inline recipes-samples/images/rpb-qt5-image.bb]: where are these apps coming from?  do they work? let's see if we include them or not. 

the one example i had in mind was qtbase-examples.

File: recipes-samples/images/rpb-qt5-image.bb (hunk 6 above)
Comment (ldts) [inline recipes-samples/images/rpb-qt5-image.bb]: I can start again but yes, there is a dependency with X11. do (libglu IIRC). I can rebuild and send you a log?

File: recipes-samples/images/rpb-qt5-image.bb (hunk 7 above)
Comment (ldts) [inline recipes-samples/images/rpb-qt5-image.bb]: these are the apps contained in the examples.
yes they do work with --platform eglfs

File: recipes-samples/images/rpb-qt5-image.bb (hunk 6 above)
Comment (ndechesne) [inline recipes-samples/images/rpb-qt5-image.bb]: hmm. i found the issue, it is a bug upstream, but in meta-qcom ;-) it's coming from 
https://github.com/ndechesne/meta-qcom/blob/master/recipes-multimedia/gstreamer/gstreamer1.0-plugins-bad_%25.bbappend

i will fix it. then we can leave gstreamer1.0-plugin-bad-meta in the image.. thanks for spotting it..

File: conf/distro/rpb-qt5.conf [hunk 8]
```diff
@@ -0,0 +1,8 @@
+require conf/distro/include/rpb.inc
+
+DISTRO_NAME = "Reference-Platform-Build-QT5"
+
+PACKAGECONFIG_append_pn-qtbase = " gles2 eglfs fontconfig kms"
+PACKAGECONFIG_append_pn-examples = " examples"
```
Comment (ndechesne) [inline conf/distro/rpb-qt5.conf]: this should be in pn-qtbase as well, pn-examples doesn't do anything.

File: conf/distro/rpb-qt5.conf (hunk 1 above)
Comment (ndechesne) [inline conf/distro/rpb-qt5.conf]: i tried to build the qt5 image with RPB/X11 distro and it failed. I think this is because gles2 and fontconfig should be enabled in rpb.inc, and we should keep eglfs and kms only here.  i will test that..

Comment (ndechesne) [inline conf/distro/rpb-qt5.conf]: yes, i confirm that we need to set gles2 and fontconfig in rpb.inc and only set eglfs and kms in rpb-qt5.conf. with that change we can build the Qt5 image for X11 and for wayland too.

Comment (ldts) [inline conf/distro/rpb-qt5.conf]: yes it is building on my side as well. should I send an updated patch or will you push your own?

File: conf/distro/rpb-qt5.conf (hunk 8 above)
Comment (ldts) [inline conf/distro/rpb-qt5.conf]: ok

**Comment by koenkooi** [comment]
what about rpb-eglfs as a name for it?

**Comment by ldts** [comment]
yep, renamed

**Comment by ldts** [comment]
anything else?

**Comment by ndechesne** [comment]
need some time for testing/trying out ;-)

**Comment by politoleo** [comment]
Hello, I'm trying to build using this configuration but it's failing with this error:

_ERROR: Nothing PROVIDES 'libglu' (but /oe-rpb/build-rpb-eglfs/conf/../../layers/openembedded-core/meta/recipes-multimedia/gstreamer/gstreamer1.0-plugins-bad_1.6.3.bb DEPENDS on or otherwise requires it)
ERROR: libglu was skipped: missing required distro feature 'x11' (not in DISTRO_FEATURES)
NOTE: Runtime target 'gstreamer1.0-plugins-bad-meta' is unbuildable, removing...
Missing or unbuildable dependency chain was: ['gstreamer1.0-plugins-bad-meta', 'libglu']
ERROR: Required build target 'rpb-qt5-image' has no buildable providers.
Missing or unbuildable dependency chain was: ['rpb-qt5-image', 'gstreamer1.0-plugins-bad-meta', 'libglu']_

Is there something obvious I'm missing?
Thank you

**Comment by ndechesne** [comment]
this error was fixed during the review, it was an issue in meta-qcom actually. the fix is this one:

https://github.com/ndechesne/meta-qcom/commit/7be7a1c3dd1a81628ed521c0630b0efb44c6de17

Note that we've tested only with master and morty, which have a 'recent' qt5. You seem to be using krogoth which has a much older qt version. if you can, you should upgrade.. we have seen issues with qt5/krogoth/eglfs and we have ignored them since we are moving the main release to morty..

**Comment by politoleo** [comment]
Thank you, I'm switching to master right now!

**Comment by ndechesne** [comment]
thanks @ldts I tried on DB410c and it worked fine. It is a bit slow , but I suspect this is the best we can do on this GPU (i was using 1080p display). I tried various qt5 examples included in the image.. 

thanks a lot for sending this out.

**Comment by ldts** [comment]
cool. thanks for your help on it!
